services/src/main/scala/cromwell/services/metadata/impl/pubsub/PubSubMetadataServiceActor.scala

```diff
@@ -0,0 +1,133 @@
+  */
+class PubSubMetadataServiceActor(serviceConfig: Config, globalConfig: Config) extends Actor with ActorLogging {
+  implicit val ec = context.dispatcher
+
+  // The auth *must* be a service account auth but it might not be named service-account.
+  val googleAuthName = serviceConfig.as[Option[String]]("auth").getOrElse("service-account")
+
+  val googleProject = serviceConfig.as[String]("project")
+  val pubSubTopicName = serviceConfig.as[Option[String]]("topic").getOrElse("cromwell-metadata")
+  val pubSubSubscriptionName = serviceConfig.as[Option[String]]("subscription")
+  val pubSubAppName = serviceConfig.as[Option[String]]("appName").getOrElse("cromwell")
+
+  val pubSubConnection = createPubSubConnection()
+  createTopicAndSubscription().failed foreach { e => throw e }
+
+  override def receive = {
+    case action: PutMetadataAction =>
+      publishMessages(action.events).failed foreach { e =>
+        log.error(e, "Failed to post metadata: " + action.events)
+      }
+    case action: PutMetadataActionAndRespond =>
+      publishMessages(action.events) onComplete {
+        case Success(_) => action.replyTo ! MetadataWriteSuccess(action.events)
+        case Failure(e) => action.replyTo ! MetadataWriteFailure(e, action.events)
+      }
+  }
+
+  protected[this] def createPubSubConnection(): GooglePubSubDAO = {
+    implicit val as = context.system
+
+    val googleConfig = GoogleConfiguration(globalConfig)
+
+    // This class requires a service account auth due to the library used
+    val googleAuth = googleConfig.auth(googleAuthName) match {
+      case Valid(a: ServiceAccountMode) => a
+      case Valid(doh) => throw new IllegalArgumentException(s"Unable to configure PubSubMetadataServiceActor: ${doh.name} was not a service account auth")
+      case Invalid(e) => throw new IllegalArgumentException("Unable to configure PubSubMetadataServiceActor: " + e.toList.mkString(", "))
+    }
+
+    val pemFile = googleAuth.fileFormat match {
+      case p: PemFileFormat => p
+      case _ => throw new IllegalArgumentException("Unable to configure PubSubMetadataServiceActor: the service account must supply a PEM file")
+    }
+
+    // The metric name is pretty useless. It's an artifact of the workbench libraries which we're not using
+    new HttpGooglePubSubDAO(pemFile.accountId, pemFile.file, pubSubAppName, googleProject, "cromwell.pubsub")
+  }
+
+  /**
+    * Creates the specified topic, and optionally the subscription if it was defined. We are ignoring the bool return
+    * type from the HttpGooglePubSubDAO return types as it is not considered an error if the topic/subscription already
+    * exist (in fact, this is a likely circumstance).
+    */
+  private def createTopicAndSubscription(): Future[Unit] = {
+    log.info("Ensuring topic " + pubSubTopicName + " exists")
+    pubSubConnection.createTopic(pubSubTopicName).flatMap(_ => createSubscription()).void
```

Comment:
@mcovarr you had asked for an `as` here on top of the `void`, what did you mean?

Reply:
ignore me, I missed the fact that's a `flatMap` and not a `map`.

Reply:
oh duh. it *was* a `map` but when i added the void I realized i had a `Future[Future[Unit]]` and it was ok because of the `Future[Unit]` return type